Need a method to take a UIKeyCommand and return whether any Web Extension command matches

eddydas opened the pull request:
#### fdb3855c7cd3ac3bba2c68998b92948e4fb3be9d
<pre>
Need a method to take a UIKeyCommand and return whether any Web Extension command matches
<a href="https://bugs.webkit.org/show_bug.cgi?id=273920">https://bugs.webkit.org/show_bug.cgi?id=273920</a>
<a href="https://rdar.apple.com/127787491">rdar://127787491</a> (Need a method to take a UIKeyCommand and return whether any Web Extension command matches (273920))

Reviewed by Brian Weinstein.

Introduces a couple of convenience methods that triggers the applicable
WebExtensionCommand based on a given UIKeyCommand.

This pattern is done for AppKit already, but lacks the UIKit counterpart.

This is needed because the existing _WKWebExtensionKeyCommand.handler is
never called because _WKWebExtensionKeyCommand itself will never be in the
responder chain. Therefore the app needs to pass the received UIKeyCommand
object back to the extensions stack for handling. The new methods in this patch
enables such handling.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext performCommandForKeyCommand:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:identifier:]):
(-[_WKWebExtensionKeyCommand performWebExtensionCommandForKeyCommand:]):
(WebKit::WebExtensionCommand::keyCommand const):

(WebKit::WebExtensionCommand::matchesKeyCommand const):
Ideally WebExtensionCommand should have a globally unique identity that can
be propagated into UIKeyCommand.propertyList. But there isn&apos;t such identity
right now. Let&apos;s just check for the input, modifierKeys, and the dev-provided command identifier for now.

(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]): Deleted.
UIKeyCommand isn&apos;t really meant to be subclassed. This subclass exists today
just to attach a _handler to the UIKeyCommand object. But when the key is
pressed, the UIKeyCommand object we will receive is actually another object
created by the system. So this _handler is actually never called.
So, let&apos;s vend the base class UIKeyCommand instead.
_WKWebExtensionKeyCommand still exists after this patch just to provide a
namespace for the designated selector -performWebExtensionCommandForKeyCommand:
to live somewhere reasonable.

(-[_WKWebExtensionKeyCommand copyWithZone:]): Deleted.
(-[_WKWebExtensionKeyCommand performWithSender:target:]): Deleted.
(-[_WKWebExtensionKeyCommand _resolvedTargetFromFirstTarget:sender:]): Deleted.
(-[_WKWebExtensionKeyCommand _performWithTarget:]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::command):
(WebKit::WebExtensionContext::performCommand):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: <a href="https://commits.webkit.org/278619@main">https://commits.webkit.org/278619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/285a2927a5d99dcad8bf5d14858b04012495157b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30427 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1815 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1489 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41638 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53224 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44064 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22754 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25386 "4 new passes 7 flakes") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55978 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26235 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1290 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49034 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27483 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48168 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28364 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7427 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -48,48 +48,29 @@
-    if (_handler)
-        _handler();
+    // To be handled by the first responder instead of this class.
+    // This method exists just to make this symbol available.
```
Should this have an `ASSERT_NOT_REACHED()` if it's never meant to be called?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2619,6 +2619,33 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 }
 
+#if TARGET_OS_IPHONE
+WebExtensionCommand* WebExtensionContext::command(UIKeyCommand *keyCommand)
```
This name could be a little bit more descriptive. `commandForKeyCommand`? `matchingCommand`?

eddydas replied:
Yeah I think so too. 

We do have these 2 in the same class though:
```
    WebExtensionCommand* command(const String& identifier);
    WebExtensionCommand* command(NSEvent *);
```
Hence me naming it the same.

I'll call it commandMatchingKeyCommand()

eddydas committed:
Need a method to take a UIKeyCommand and return whether any Web Extension command matches
https://bugs.webkit.org/show_bug.cgi?id=273920
rdar://127787491 (Need a method to take a UIKeyCommand and return whether any Web Extension command matches (273920))

Reviewed by Brian Weinstein.

Introduces a couple of convenience methods that triggers the applicable
WebExtensionCommand based on a given UIKeyCommand.

This pattern is done for AppKit already, but lacks the UIKit counterpart.

This is needed because the existing _WKWebExtensionKeyCommand.handler is
never called because _WKWebExtensionKeyCommand itself will never be in the
responder chain. Therefore the app needs to pass the received UIKeyCommand
object back to the extensions stack for handling. The new methods in this patch
enables such handling.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext performCommandForKeyCommand:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:identifier:]):
(-[_WKWebExtensionKeyCommand performWebExtensionCommandForKeyCommand:]):
(WebKit::WebExtensionCommand::keyCommand const):

(WebKit::WebExtensionCommand::matchesKeyCommand const):
Ideally WebExtensionCommand should have a globally unique identity that can
be propagated into UIKeyCommand.propertyList. But there isn't such identity
right now. Let's just check for the input, modifierKeys, and the dev-provided command identifier for now.

(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]): Deleted.
UIKeyCommand isn't really meant to be subclassed. This subclass exists today
just to attach a _handler to the UIKeyCommand object. But when the key is
pressed, the UIKeyCommand object we will receive is actually another object
created by the system. So this _handler is actually never called.
So, let's vend the base class UIKeyCommand instead.
_WKWebExtensionKeyCommand still exists after this patch just to provide a
namespace for the designated selector -performWebExtensionCommandForKeyCommand:
to live somewhere reasonable.

(-[_WKWebExtensionKeyCommand copyWithZone:]): Deleted.
(-[_WKWebExtensionKeyCommand performWithSender:target:]): Deleted.
(-[_WKWebExtensionKeyCommand _resolvedTargetFromFirstTarget:sender:]): Deleted.
(-[_WKWebExtensionKeyCommand _performWithTarget:]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::command):
(WebKit::WebExtensionContext::performCommand):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: https://commits.webkit.org/278619@main
